For #9599: Ability to edit top site URLs

pepve opened the pull request:
* ~~Add functionality to edit both the title and the URL of a top site simultaneously~~
* ~~Deprecate functionality to edit only the title of a top site~~

Replace the functionality to rename top sites with functionality to update both the title and the URL.

### Pull Request checklist
<!-- Before submitting the PR, please address each item -->
- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

gabrielluong commented on components/feature/top-sites/src/main/java/mozilla/components/feature/top/sites/DefaultTopSitesStorage.kt:
```diff
@@ -76,6 +76,16 @@ class DefaultTopSitesStorage(
         }
     }
 
+    override fun editTopSite(topSite: TopSite, title: String, url: String) {
```
Let's remove renameTopSite since I don't imagine any consumers of it after this. Can we also call it `updateTopSite` instead of `editTopSite`?

pepve replied:
Done and done!

gabrielluong commented on components/feature/top-sites/src/main/java/mozilla/components/feature/top/sites/PinnedSiteStorage.kt:
```diff
@@ -88,6 +88,20 @@ class PinnedSiteStorage(context: Context) {
+     * @param title The new title for the top site.
+     * @param url The new url for the top site.
+     */
+    suspend fun editPinnedSite(site: TopSite, title: String, url: String) = withContext(IO) {
```
Can we also make sure there is sufficient unit tests for this in PinnedSiteStorageTest.

pepve replied:
Yes we can, thanks, I missed that!

gabrielluong commented:
I'm planning to look at how the edit bookmarks (use case if there is one?) works tomorrow to investigate how it works - looking for clues on how/if it handles empty URL provided and perhaps malformed URLs. Will also look at other browser to see how much we will need to dive into this. Feel free to also investigate this, I just want to make sure we cover our bases. Thanks again for contributing!

pepve commented:
I just rebased on the latest master. I also added a commit with a test to improve the (measured) test coverage, please review, I'm not entirely sure of the `mockDao` function I came up with.

About validation, can we pick that up in a separate issue? From the AC perspective it was already possible to add any random string as a URL. And the new update logic doesn't change that. The new issue will then of course block [the Fenix issue I'm doing this for: #17636](https://github.com/mozilla-mobile/fenix/issues/17636).

gabrielluong commented on components/feature/top-sites/src/test/java/mozilla/components/feature/top/sites/PinnedSitesStorageTest.kt:
```diff
@@ -0,0 +1,127 @@
+class PinnedSitesStorageTest {
+
+    @Test
+    fun `add all default sites`() = runBlocking {
```
Perhaps we can simply use the `addAllPinnedSites` as the test name and the others.

pepve replied:
Done!

gabrielluong commented on components/feature/top-sites/src/test/java/mozilla/components/feature/top/sites/PinnedSitesStorageTest.kt:
```diff
@@ -0,0 +1,127 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
This is nice!

pepve replied:
Thank you!

pepve commented on components/feature/top-sites/src/androidTest/java/mozilla/components/feature/top/sites/OnDevicePinnedSitesStorageTest.kt:
```diff
@@ -28,7 +28,7 @@ import java.util.concurrent.Executors
 
 @Suppress("LargeClass")
-class PinnedSitesStorageTest {
+class OnDevicePinnedSitesStorageTest {
```
I renamed this because now there is also a _regular_ test. And in most cases where a class is tested in both `test` and `androidTest`, the test in `androidTest` gets prefixed with `OnDevice`. I found these examples:

`components/browser/icons/src/androidTest/java/mozilla/components/browser/icons/OnDeviceBrowserIconsTest.kt`
`components/browser/icons/src/test/java/mozilla/components/browser/icons/BrowserIconsTest.kt`

`components/support/images/src/androidTest/java/mozilla/components/support/images/decoder/OnDeviceAndroidImageDecoderTest.kt`
`components/support/images/src/test/java/mozilla/components/support/images/decoder/AndroidImageDecoderTest.kt`

`components/feature/sitepermissions/src/androidTest/java/mozilla/components/feature/sitepermissions/db/OnDeviceSitePermissionsStorageTest.kt`
`components/feature/sitepermissions/src/test/java/mozilla/components/feature/sitepermissions/SitePermissionsStorageTest.kt`

`components/feature/prompts/src/androidTest/java/mozilla/components/feature/prompts/file/OnDeviceFilePickerTest.kt`
`components/feature/prompts/src/test/java/mozilla/components/feature/prompts/file/FilePickerTest.kt`

`components/feature/downloads/src/androidTest/java/mozilla/components/feature/downloads/OnDeviceDownloadStorageTest.kt`
`components/feature/downloads/src/test/java/mozilla/components/feature/downloads/DownloadStorageTest.kt`

And only one example to the contrary:

`components/browser/icons/src/androidTest/java/mozilla/components/browser/icons/decoder/ICOIconDecoderTest.kt`
`components/browser/icons/src/test/java/mozilla/components/browser/icons/decoder/ICOIconDecoderTest.kt`

Of course I will happily oblige and not rename this class. But I want to make sure you know my reasons for proposing to rename it. Please let me know.

gabrielluong replied:
Gotcha, I think this is good. Thanks for the details!

pepve commented on components/feature/top-sites/src/main/java/mozilla/components/feature/top/sites/PinnedSiteStorage.kt:
```diff
@@ -16,6 +17,9 @@ import mozilla.components.feature.top.sites.db.toPinnedSite
  */
 class PinnedSiteStorage(context: Context) {
 
+    @VisibleForTesting
+    internal var currentTimeMillis: () -> Long = { System.currentTimeMillis() }
```
I use it to inject a constant time from the tests (`add al default sites` and `add pinned site`) so that I can use `verify` without much effort. What would be a good alternative?

gabrielluong replied:
I think we can just check for anyLong() as an alternative.

pepve replied:
I tried that but I think it doesn't work because the long is wrapped in an object. I think this works:
`verify(obj).method(eq("foo"), anyLong())`
But this doesn't:
`verify(obj).method(SomeClass(eq("foo"), anyLong()))`
And our case matches the latter. But I don't really have a lot of experience in this area so please correct me if I'm missing something. The real problem I think we're having is that there is no consistent way to mock time used throughout the project. Code that does straight calls to `System.currentTimeMillis()` is just not very unit-testable. Perhaps we should make an issue about this?

pepve commented on components/support/migration/src/test/java/mozilla/components/support/migration/FennecMigratorTest.kt:
```diff
@@ -312,6 +312,8 @@ class FennecMigratorTest {
             "Internet for people, not profit — Mozilla",
             "https://www.mozilla.org/en-US/"
         )
+
+        Unit
```
This change https://github.com/mozilla-mobile/android-components/pull/9600/files#diff-d3f84dd311db914486892c2b2d7390bf9a56ef23f4e5b59b95f7fbac3a025d8aL60-R64 changed the return type for `PinnedSiteStorage.addPinnedSite` which is called here.

pepve replied:
Undid the change.

pepve commented on docs/changelog.md:
```diff
@@ -49,6 +49,9 @@ permalink: /changelog/
   * Add `threadNamePrefix` parameter to `Store` to give threads created by the `Store` a specific name.
 
+* **feature-top-sites**
+  * ⚠️ **This is a breaking change**: Replace `TopSitesUseCases.renameTopSites` (which only updates the title) with `TopSitesUseCases.updateTopSites` (which also updates the url). [#9599](https://github.com/mozilla-mobile/android-components/issues/9599)
```
Done!

pepve commented on components/feature/top-sites/src/main/java/mozilla/components/feature/top/sites/PinnedSiteStorage.kt:
```diff
@@ -55,9 +59,9 @@ class PinnedSiteStorage(context: Context) {
+                createdAt = currentTimeMillis()
             )
-            entity.id = pinnedSiteDao.insertPinnedSite(entity)
+            pinnedSiteDao.insertPinnedSite(entity)
```
Okay I'm undoing this change. I opened #9708 for it. See also my comment here: https://github.com/mozilla-mobile/android-components/pull/9600/files#diff-3a51f71b988baa050dc7351cdbdb1d06369d75afe7fcb0951091d3d642d6318cR56-R59

pepve commented:
@gabrielluong is there anything else I can do to help progress this?

pepve commented:
Thank you!
